alphadia/workflow/optimization.py

```diff
@@ -400,13 +422,63 @@ def __init__(
         self.estimator_group_name = "precursor"
         self.estimator_name = "rt"
         self.feature_name = "precursor_proportion_detected"
+        self.maximal_decrease = workflow.config["optimization"]["rt_error"][
+            "maximal_decrease"
+        ]
+        self.minimum_proportion_of_maximum = workflow.config["optimization"][
+            "rt_error"
+        ]["minimum_proportion_of_maximum"]
         super().__init__(initial_parameter, workflow, reporter)
 
     def _get_feature_value(
         self, precursors_df: pd.DataFrame, fragments_df: pd.DataFrame
     ):
         return len(precursors_df) / self.workflow.optlock.total_elution_groups
 
+    def _check_convergence(self):
+        """Optimization should stop if continued narrowing of the parameter is not improving the feature value.
+        This function checks if the previous rounds of optimization have led to a meaningful improvement in the feature value.
+        If so, it continues optimization and appends the proposed new parameter to the list of parameters. If not, it stops optimization and sets the optimal parameter attribute.
+
+        Notes
+        -----
+            Because the check for an increase in feature value requires two previous rounds, the function will also initialize for another round of optimization if there have been fewer than 3 rounds.
+
+
+        """
+        if len(self.history_df) <= 2:
+            return False
+
+        min_steps_reached = (
+            self.num_prev_optimizations
+            >= self.workflow.config["calibration"]["min_steps"]
+        )
+
+        feature_substantially_decreased = (
+            self.history_df[self.feature_name].iloc[-1]
+            < self.maximal_decrease * self.history_df[self.feature_name].iloc[-2]
+            and self.history_df[self.feature_name].iloc[-1]
+            < self.maximal_decrease * self.history_df[self.feature_name].iloc[-3]
```

Comment:
if you go to `-3` here, should the check in l. 449 also be `<=3`?

Reply:
No, because that will only stop returning False when we have at least three rows in the dataframe, at which point the -3 index should not pose any problems.

Reply:
But I've now made it < 3 instead of <= 2 for clarity.